warehouse/accounts/views.py

```diff
@@ -252,7 +252,9 @@ def request_password_reset(request, _form_class=RequestPasswordResetForm):
     form = _form_class(request.POST, user_service=user_service)
 
     if request.method == "POST" and form.validate():
-        user = user_service.get_user_by_username(form.username.data)
+        user = user_service.get_user_by_username(form.username_or_email.data)
+        if user is None:
+            user = user_service.get_user_by_email(form.username_or_email.data)
```

Comment:
Looks like we'll need to add some tests around this conditional. We also don't want to call `send_password_reset_email` if `user` is `None`, so we'll need another conditional there as well (and tests for that condition).

Reply:
Were we not calling `send_password_reset_email` with `None` when a user with that username was not found? That's why I kept it that way, I assumed that the function did a noop when None was passed in as a user.

How would I go about testing that last condition? As in, testing that nothing is done if a user is not found.

Reply:
The `UsernameMixin` was validating that the username existed, so previously the form validation would fail. This won't happen here anymore. (BTW, `RequestPasswordResetForm` is still inheriting from `UsernameMixin`, which it shouldn't be anymore)

To test the last condition, I would monkeypatch `send_password_reset_email` the same way we do in other tests, and just assert that it is not called.

Reply:
Yep, I wanted to remove `UsernameMixin` from the form, but some tests were relying on its validation of the username and I didn't know if I should remove them.
I guess instead of verifying that the username exists I'll try to change the tests to check if the result is the `{'n_hours': n_hours}` dict or `None`

As for the last condition, I'm adding another test to assert `send_password_reset_email` is not being called.

Funny how all this `pretend` functions work :) I still need to learn more about testing frameworks and modules

Reply:
Yeah, we should remove it and update the tests -- we don't need the `username` field on this form anymore.

Hmm, we might have to tweak this a little bit so that `n_hours` can still be passed to the template even if `send_password_reset_email` is not called. We can get it from the `TokenService` like so:

```python
token_service = request.find_service(ITokenService, name='password')
n_hours = token_service.max_age // 60 // 60
```
Let's just use this instead of what `send_password_reset_email` sends back.